Algorithm tests

Rod1122 opened the pull request:
- Modified base algorithm to store the train set in a class variable along with the unique users and items present in the train set.
- Modified base predictor to remove any users/items from the prediction results that are not known from the training stage.
- Modified base recommender to remove any users from the recommendation results that are not known from the training stage.
- Updated all derived predictors/recommenders to satisfy previous changes.
- Added tests for all predictors/recommenders to obey their respective standard interfaces.
- Added test for algorithm creation (with parameters).
- Moved model factory test.

Commit:
Algorithm tests

- Modified base algorithm to store the train set in a class variable along with the unique users and items present in the train set.
- Modified base predictor to remove any users/items from the prediction results that are not known from the training stage.
- Modified base recommender to remove any users from the recommendation results that are not known from the training stage.
- Updated all derived predictors/recommenders to satisfy previous changes.
- Added tests for all predictors/recommenders to obey their respective standard interfaces.
- Added test for algorithm creation (with parameters).
- Moved model factory test.

mendoxei commented on src/fairreckitlib/model/algorithms/elliot/elliot_params.py:
```diff
@@ -43,7 +42,7 @@ def create_params_knn() -> ConfigParameters:
 
     params = ConfigParameters()
-    params.add_value('neighbours', int, 40, (1, 50))
+    params.add_value('neighbors', int, 40, (1, 50))
```
🤔 we should probably decide on American/British English as a group

Rod1122 replied:
yea haha, I changed most of the param names in elliot to make them a little more descriptive than things like 'lr'

mendoxei commented on tests/test_algorithms.py:
```diff
@@ -0,0 +1,327 @@
+    assert bool(model_factory.get_factory(TYPE_PREDICTION)), 'missing prediction models.'
+    assert bool(model_factory.get_factory(TYPE_RECOMMENDATION)), 'missing recommender models.'
+
+    for _, model_type in enumerate(model_factory.get_available_names()):
```
Why not for model_type in model_factory.get_available_names() tho?

Rod1122 replied:
right, all those for loops could be simplified then. I have been working with enumerate too much

mendoxei replied:
I mean we can discuss it with the TA, because using enumerate in this manner consistently seems pretty fine to me too. That's why I didn't bring it up before :p

Rod1122 replied:
I have had the pylinter complain about it in the past, so I have been using it too often ever since.
Actually removing enumerate did reduce one of the too many local variables.

mendoxei reviewed (approved):
Solid as always

TheMinefreak23 commented on src/fairreckitlib/model/algorithms/elliot/elliot_recommender.py:
```diff
@@ -32,17 +32,17 @@ def __init__(self, name: str, params: Dict[str, Any], **kwargs):
+    def on_train(self) -> None:
         """Train the elliot model not supported."""
         # not used, training is done by running the framework
         raise NotImplementedError()
```
If a method is not supposed to be implemented, it shouldn't raise a NotImplementedError.
https://docs.python.org/3/library/exceptions.html#NotImplementedError

Rod1122 replied:
good point, ill change them to RuntimeError's as well

TheMinefreak23 reviewed (changes requested):
Tests seem to work fine, and I trust you'll fix the linting issue soon :)

Commit:
Refactoring algorithm tests

Modified the predictor/recommender interface tests into smaller chunks of code.
Modified Elliot recommender to return RuntimeError's instead of Notimplemented.
Added test to check if the subset dataframe is sufficient to run algorithm tests on.
Added extra assertions to check if the header in a returned dataframe is correct.
Added extra assertion to check prediction batching for multiple users.
